### PR TITLE
`ultralytics 8.4.35` NaN training recovery and dataset cache improvements

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary
- recover NaN training runs from `last_good.pt` instead of retrying corrupted `last.pt`
- avoid writing empty detection caches and surface corrupt label reasons in the final error
- accept dataset directories directly in detection dataset checks
- rebuild cached NDJSON conversions when expected split directories are missing
- throttle repeated `Platform: Model not found` warnings during a run

## Validation
- `python -m py_compile ultralytics/engine/trainer.py ultralytics/data/dataset.py ultralytics/data/utils.py ultralytics/data/converter.py ultralytics/utils/callbacks/platform.py tests/test_engine.py`
- `python -m pytest tests/test_engine.py -k nan_recovery -q`

Companion portal PR updates worker GPU/process cleanup and training start validation.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
This PR improves dataset handling and training reliability by tightening cache validation, surfacing clearer dataset errors, supporting dataset directory inputs, and preventing invalid model checkpoints from being saved 🚀

### 📊 Key Changes
- Improved NDJSON-to-YOLO dataset conversion caching:
  - Cached conversions are now reused only if the dataset hash matches **and** the expected image/label split folders actually exist (`train`, `val`, `test`) 📁
- Safer dataset label cache writing:
  - Label cache files are now saved only when valid labels are found, avoiding empty or misleading cache files ✅
- Clearer dataset error reporting:
  - When no valid images are found, the error now includes the actual warning/details collected during validation, making debugging much easier 🛠️
- Better dataset path handling:
  - Detection dataset checks now accept a dataset **directory** directly and automatically locate the dataset YAML file inside it 🔎
- More robust checkpoint saving during training:
  - Training now checks whether EMA weights contain `NaN` or `Inf` before saving checkpoints
  - If invalid values are detected, checkpoint saving is skipped and a warning is logged instead of writing a bad model file ⚠️
- Callback behavior refined:
  - `on_model_save` callbacks now run only when a model checkpoint is successfully saved
- Stronger NaN recovery logic:
  - If training corruption is detected and recovery is needed, the code now raises a clear error when `last.pt` is missing instead of silently continuing in a broken state 🧯
- Package version bumped:
  - Updated Ultralytics version from `8.4.34` to `8.4.35` 🆕

### 🎯 Purpose & Impact
- Reduces confusing dataset conversion/cache issues by ensuring cached results are only reused when the expected files are really present
- Helps users diagnose bad or misformatted datasets faster with more informative error messages
- Makes dataset loading more convenient by allowing users to point directly to a dataset folder instead of manually specifying its YAML file
- Prevents corrupted or unstable training runs from producing invalid checkpoint files, improving trust in saved models
- Makes recovery from NaN-related training failures safer and more explicit, which is especially useful for long or distributed training jobs
- Overall, this update improves stability, debugging, and usability for both new users and advanced training workflows 🎯

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help streamline review; an engineer will assist soon. Please review the checklist below to help ensure smooth integration. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -662,6 +663,8 @@ def save_model(self):
         # Save checkpoints
         self.wdir.mkdir(parents=True, exist_ok=True)  # ensure weights directory exists
         self.last.write_bytes(serialized_ckpt)  # save last.pt
+        if all(torch.isfinite(v).all() for v in ema.state_dict().values() if isinstance(v, torch.Tensor)):
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: ⚠️ **HIGH**: `last_good.pt` is only gated on EMA tensor finiteness, but `_handle_nan_recovery()` later restores the optimizer and scaler from the same checkpoint. If the optimizer moments or GradScaler have already gone non-finite while EMA is still finite, recovery will reload the bad state and can fall straight back into NaNs. Please either validate those states before updating `last_good.pt`, or avoid restoring them during NaN recovery.

File: ultralytics/data/converter.py [hunk 2]
```diff
@@ -843,7 +843,12 @@ async def convert_ndjson_to_yolo(ndjson_path: str | Path, output_path: str | Pat
     yaml_path = dataset_dir / "data.yaml"
     if yaml_path.is_file():
         try:
-            if YAML.load(yaml_path).get("hash") == _hash:
+            cached = YAML.load(yaml_path)
+            if cached.get("hash") == _hash and all(
+                (dataset_dir / cached[split]).is_dir() and (dataset_dir / "labels" / split).is_dir()
```
Comment (UltralyticsAssistant) [inline ultralytics/data/converter.py:848]: 💡 **MEDIUM**: This now requires a `labels/<split>` directory for every split listed in the cached YAML. That will invalidate otherwise-correct caches for datasets that intentionally have an image-only `test` split, forcing reconversion every time. Can we limit the label-directory check to annotated splits, or at least exclude `test` here?

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good overall direction: preferring `last_good.pt` should make recovery more reliable, and the dataset error reporting is clearer. I found two edge cases worth addressing before merge: `last_good.pt` can still capture corrupt optimizer/scaler state, and the new NDJSON cache validation is too strict for image-only test splits.

💬 Posted 2 inline comments

**Comment by UltralyticsAssistant** [comment]
Thrilled to see this merged, @glenn-jocher — thank you for another thoughtful reliability upgrade to Ultralytics! 🚀

As Marie Curie said, *“Nothing in life is to be feared, it is only to be understood.”* This PR really brings that spirit to life by making dataset issues easier to understand, cache behavior safer, and checkpoint saving more trustworthy. The result is a smoother, more dependable training experience for everyone building with Ultralytics.

Really appreciate the care behind these stability and usability improvements — they make a meaningful difference for the whole community 🙌
